crates/authifier/src/config/sso.rs

```diff
@@ -0,0 +1,274 @@
+use std::{
+    borrow::Borrow,
+    collections::{HashMap, HashSet},
+    hash::{Hash, Hasher},
+    ops::Deref,
+    str::FromStr,
+};
+
+use reqwest::Url;
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "lowercase", tag = "type")]
+pub enum Endpoints {
+    Discoverable,
+    Manual {
+        authorization: Box<Url>,
+        token: Box<Url>,
+        userinfo: Box<Url>,
+    },
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "lowercase", tag = "type")]
+pub enum Credentials {
+    None {
+        client_id: String,
+    },
+    Basic {
+        client_id: String,
+        client_secret: String,
+    },
+    Post {
+        client_id: String,
+        client_secret: String,
+    },
+}
+
+impl Credentials {
+    pub fn client_id(&self) -> &str {
+        match self {
+            Credentials::None { client_id }
+            | Credentials::Basic { client_id, .. }
+            | Credentials::Post { client_id, .. } => client_id,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Clone, PartialEq, Eq, Hash)]
+#[serde(rename_all = "lowercase")]
+pub enum Claim {
+    Id,
+    Username,
+    Picture,
+    Email,
+    Custom(String),
+}
+
+#[derive(Clone, Debug)]
+pub struct IdProvider {
+    pub id: String,
+
+    pub issuer: Url,
+    pub name: Option<String>,
+    pub icon: Option<Url>,
+
+    pub scopes: Vec<String>,
+    pub endpoints: Endpoints,
+    pub credentials: Credentials,
+    pub claims: HashMap<Claim, String>,
+
+    pub code_challenge: bool,
+}
+
+impl Borrow<str> for IdProvider {
+    fn borrow(&self) -> &str {
+        &self.id
+    }
+}
+
+impl PartialEq for IdProvider {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl Eq for IdProvider {}
+
+impl Hash for IdProvider {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: Hasher,
+    {
+        self.id.hash(state);
+    }
+}
+
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct SSO(HashSet<IdProvider>);
```

Comment:
Is there a reason this isnt a `HashMap`? would remove the need for a custom `Serialize`, `Deserialize` and `Hash` impl

```suggestion
pub struct SSO(pub HashSet<IdProvider>);
```